### ISSUE TITLE
How to lower the total needed shared memory?

### ISSUE THREAD

**Issue by yangarbiter** [botonchou/libdnn#8]
Hi Chou,

I am encountering the following problem when trying to run convolutional neural network.

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  [Error] In function "getSuitableShmConfig" (at src/cnn-utility.cu:418): Exceeds maximum shared memory available. (49152 bytes)
kernel = (100, 92), grids = (8, 8, 40), threads = (4, 4, 1)  => 75940 bytes of shared memory needed.
```

What is the standard way to fix this problem if I really don't have that much memory? (ex. lowering thread count)

Thanks in advance.

**Comment by yangarbiter** [comment]
Actually the kernel size 23*14 is able to be run on my computer using theano. I am still wondering about what Caffe have done to get the 7% error rate. I can only achieve around 15% error rate with Lenet implemented by theano.

BTW, I found out the the nn-init would crash without any error message when the format of input training data is wrong. (I accidentally start my array from zero)

**Comment by botonchou** [comment]
Can you please tell me the kernel size you choose in your LeNet ?

Okay, I'll add a guard clause at line 145 in **src/data-io.cpp**:
``` c++
size_t j = stof(token.substr(0, pos)) - 1;
```
Thanks !!

**Comment by yangarbiter** [comment]
The filter size is 23x14 and 14x12
number of kernel is 40 and 100

I ran it on GTX780, the implementation is pretty much like the code in the following link. http://deeplearning.net/tutorial/lenet.html
I think the kernel size in libdnn is the same as LeNetConvPoolLayer.filter_shape in theano, right?

**Comment by botonchou** [comment]
Yes, kernel is filter.

But the error message you mentioned before `(kernel = (100, 92), grids = (8, 8, 40), threads = (4, 4, 1)) ` is the the way you think. XD
You see at line 772 in file **src/feature-transform.cpp** 
```c++
Z[i] += convn_2(rot180(iImgs[i][k]), Y[k], this->get_output_img_size());
```
and line 442 in file **src/cnn-utility.cu**
```c++
mat convn_2(const mat& data, const mat& kernels, SIZE k) {
...
size_t SHM_SIZE = getSuitableShmConfig(grids, threads, k.m, k.n);
...
return output;
```
You can see that the third argument in `convn_2` is `this->get_output_img_size()`, which is later passed to `getSuitableShmConfig` as variable `k`. And the size of output image in the first ConvolutionalLayer is [100, 92] \( = [122, 105] - [23, 14] + [1,1] \)

Well, that doesn't means I won't fix this issue.
I'm currently refactoring src/cnn-utility.cu and try to make it more friendly, readable, and faster. (it's 50% already 50% faster now). I'll change the runtime_error message to a warning message in the next release.
But before that, I still suggest you use a smaller image by scaling down to 64 x 64, and try a smaller kernel.

I already heard that it's possible to achieve 95% in-sample accuracy and 92% out-of-sample accuracy using **3** layers of ConvolutionalLayer and SubsamplingLayer with 20% dropout. : )
